File: fragments/solana_program_round/solana_round_interface.ts
import {
  AccountRole,
  Address,
  appendTransactionMessageInstruction,
  assertAccountExists,
  Decoder,
  fetchEncodedAccount,
  getAddressDecoder,
  getOptionDecoder,
  getStructDecoder,
  getU64Decoder,
  getU64Encoder,
  KeyPairSigner,
  offsetDecoder,
  Option,
} from "@solana/kit";
import { getInstructionDiscriminator, getPda } from "../solana_program/solana_program_utils";
import { SYSTEM_PROGRAM_ADDRESS } from "@solana-program/system";
import { Buffer } from "node:buffer";
import { initRpcClient } from "../solana_rpc/solana_rpc_utils";
import {
  createBaseTxWithFeePayerAndLifetime,
  signAndSendTransaction,
} from "../solana_transaction/solana_transaction_utils";

export const initialiseRound = async (
  authority: KeyPairSigner,
  programAddress: Address,
  startSlot: bigint,
) => {
  const discriminator = getInstructionDiscriminator("initialise_round", "round");
  const payer = authority.address;
  const pda = await getPda(payer, programAddress, "round");
  const baseTx = await createBaseTxWithFeePayerAndLifetime(payer);

  const tx = appendTransactionMessageInstruction({
    programAddress,
    data: Buffer.concat([discriminator, Buffer.from(getU64Encoder().encode(startSlot))]),
    accounts: [
      { address: pda, role: AccountRole.WRITABLE },
      { address: payer, role: AccountRole.WRITABLE_SIGNER },
      { address: SYSTEM_PROGRAM_ADDRESS, role: AccountRole.READONLY },
    ],
  }, baseTx);

  const sig = await signAndSendTransaction(tx, authority.keyPair);

  return sig;
};

export const getRoundAccount = async (authority: Address, programAddress: Address) => {
  const client = initRpcClient();
  const pda = await getPda(authority, programAddress, "round");
  const account = await fetchEncodedAccount(client, pda, {
    commitment: "confirmed",
    abortSignal: AbortSignal.timeout(5000),
  });

  if (!account.exists) {
    throw new Error(`Account ${pda} does not exist`);
  }

  assertAccountExists(account);

  const decoded = roundAccountDecoder.decode(account.data);

  return decoded;
};

export const activateRound = async (payer: KeyPairSigner, programAddress: Address, authority: Address) => {
  const discriminator = getInstructionDiscriminator("activate_round", "round");
  const pda = await getPda(authority, programAddress, "round");
  const baseTx = await createBaseTxWithFeePayerAndLifetime(payer.address);

  const tx = appendTransactionMessageInstruction({
    programAddress,
    data: discriminator,
    accounts: [
      { address: pda, role: AccountRole.WRITABLE },
      { address: payer.address, role: AccountRole.WRITABLE_SIGNER },
    ],
  }, baseTx);

  const sig = await signAndSendTransaction(tx, payer.keyPair);

  return sig;
};

export const completeRound = async (authority: KeyPairSigner, programAddress: Address) => {
  const discriminator = getInstructionDiscriminator("complete_round", "round");
  const pda = await getPda(authority.address, programAddress, "round");
  const baseTx = await createBaseTxWithFeePayerAndLifetime(authority.address);

  const tx = appendTransactionMessageInstruction({
    programAddress,
    data: discriminator,
    accounts: [
      { address: pda, role: AccountRole.WRITABLE },
      { address: authority.address, role: AccountRole.WRITABLE_SIGNER },
    ],
  }, baseTx);

  const sig = await signAndSendTransaction(tx, authority.keyPair);

  return sig;
};

const roundAccountDecoder: Decoder<{
  start_slot: bigint;
  authority: Address;
  activated_at: Option<bigint>;
  activated_by: Option<Address>;
  completed_at: Option<bigint>;
}> = offsetDecoder(
  getStructDecoder([
    ["start_slot", getU64Decoder()],
    ["authority", getAddressDecoder()],
    ["activated_at", getOptionDecoder(getU64Decoder())],
    ["activated_by", getOptionDecoder(getAddressDecoder())],
    ["completed_at", getOptionDecoder(getU64Decoder())],
  ]),
  {
    preOffset: ({ wrapBytes }) => wrapBytes(8), // removes the discriminator from the account data
  },
);

File: fragments/solana_program_round/solana_round_interface.go
package solana_program_round

import (
	"context"
	"encoding/binary"
	solana_program "ff/solana_program"
	"ff/solana_rpc"
	"ff/solana_transaction"
	"fmt"

	"github.com/gagliardetto/solana-go"
	"github.com/gagliardetto/solana-go/rpc"
)

type RoundAccount struct {
	StartSlot   uint64
	ActivatedAt *uint64
	ActivatedBy *solana.PublicKey
	Authority   solana.PublicKey
	CompletedAt *uint64
}

func InitialiseRound(authority solana.PrivateKey, programId solana.PublicKey, startSlot uint64) (solana.Signature, error) {
	client := solana_rpc.InitRpcClient()
	pda := solana_program.GetProgramDerivedAddress(authority.PublicKey(), programId, "round")
	instr_discriminator, err := solana_program.GetInstructionDiscriminator("initialise_round", "round")
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to get instruction discriminator: %v", err)
	}

	startSlotBytes := make([]byte, 8)
	binary.LittleEndian.PutUint64(startSlotBytes, startSlot)
	data := append(instr_discriminator, startSlotBytes...)

	instr := solana.NewInstruction(
		programId,
		solana.AccountMetaSlice{
			&solana.AccountMeta{
				PublicKey:  pda,
				IsSigner:   false,
				IsWritable: true,
			},
			&solana.AccountMeta{
				PublicKey:  authority.PublicKey(),
				IsSigner:   true,
				IsWritable: true,
			},
			&solana.AccountMeta{
				PublicKey:  solana.SystemProgramID,
				IsSigner:   false,
				IsWritable: false,
			},
		},
		data,
	)

	tx, err := solana_transaction.CreateTxWithFeePayerAndLifetime(authority, instr)
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to create transaction: %v", err)
	}

	sig, err := client.SendTransactionWithOpts(context.Background(), tx, rpc.TransactionOpts{PreflightCommitment: rpc.CommitmentConfirmed})
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to send transaction: %v", err)
	}

	return sig, nil
}

func GetRoundAccount(authority solana.PublicKey, programId solana.PublicKey) (*RoundAccount, error) {
	client := solana_rpc.InitRpcClient()
	pda := solana_program.GetProgramDerivedAddress(authority, programId, "round")
	opts := &rpc.GetAccountInfoOpts{
		Encoding:   solana.EncodingBase64,
		Commitment: rpc.CommitmentConfirmed,
	}
	res, err := client.GetAccountInfoWithOpts(context.Background(), pda, opts)
	if err != nil {
		return nil, fmt.Errorf("failed to get round account info: %v", err)
	}
	if res.Value == nil {
		return nil, fmt.Errorf("round account not found")
	}

	data := res.Value.Data.GetBinary()
	cursor := data[8:] // Skip the 8-byte discriminator

	startSlot := binary.LittleEndian.Uint64(cursor[:8])
	cursor = cursor[8:]

	var roundOwner solana.PublicKey
	copy(roundOwner[:], cursor[:32])
	cursor = cursor[32:]

	activatedAtFlag := cursor[0]
	cursor = cursor[1:]
	var activatedAt *uint64
	if activatedAtFlag != 0 {
		val := binary.LittleEndian.Uint64(cursor[:8])
		activatedAt = &val
		cursor = cursor[8:]
	}

	activatedByFlag := cursor[0]
	cursor = cursor[1:]
	var activatedBy *solana.PublicKey
	if activatedByFlag != 0 {
		var val solana.PublicKey
		copy(val[:], cursor[:32])
		cursor = cursor[32:]
		activatedBy = &val
	}

	completedAtFlag := cursor[0]
	cursor = cursor[1:]
	var completedAt *uint64
	if completedAtFlag != 0 {
		val := binary.LittleEndian.Uint64(cursor[:8])
		completedAt = &val
	}

	return &RoundAccount{
		StartSlot:   startSlot,
		Authority:   roundOwner,
		ActivatedAt: activatedAt,
		ActivatedBy: activatedBy,
		CompletedAt: completedAt,
	}, nil
}

func ActivateRound(payer solana.PrivateKey, programId solana.PublicKey, authority solana.PublicKey) (solana.Signature, error) {
	client := solana_rpc.InitRpcClient()
	pda := solana_program.GetProgramDerivedAddress(authority, programId, "round")
	instr_discriminator, err := solana_program.GetInstructionDiscriminator("activate_round", "round")
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to get instruction discriminator: %v", err)
	}

	instr := solana.NewInstruction(
		programId,
		solana.AccountMetaSlice{
			&solana.AccountMeta{
				PublicKey:  pda,
				IsSigner:   false,
				IsWritable: true,
			},
			&solana.AccountMeta{
				PublicKey:  payer.PublicKey(),
				IsSigner:   true,
				IsWritable: true,
			},
		},
		instr_discriminator,
	)

	tx, err := solana_transaction.CreateTxWithFeePayerAndLifetime(payer, instr)
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to create transaction: %v", err)
	}

	sig, err := client.SendTransactionWithOpts(context.Background(), tx, rpc.TransactionOpts{PreflightCommitment: rpc.CommitmentConfirmed})
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to send transaction: %v", err)
	}

	return sig, nil
}

func CompleteRound(authority solana.PrivateKey, programId solana.PublicKey) (solana.Signature, error) {
	client := solana_rpc.InitRpcClient()
	pda := solana_program.GetProgramDerivedAddress(authority.PublicKey(), programId, "round")
	instr_discriminator, err := solana_program.GetInstructionDiscriminator("complete_round", "round")
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to get instruction discriminator: %v", err)
	}

	instr := solana.NewInstruction(
		programId,
		solana.AccountMetaSlice{
			&solana.AccountMeta{
				PublicKey:  pda,
				IsSigner:   false,
				IsWritable: true,
			},
			&solana.AccountMeta{
				PublicKey:  authority.PublicKey(),
				IsSigner:   true,
				IsWritable: true,
			},
		},
		instr_discriminator,
	)

	tx, err := solana_transaction.CreateTxWithFeePayerAndLifetime(authority, instr)
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to create transaction: %v", err)
	}

	sig, err := client.SendTransactionWithOpts(context.Background(), tx, rpc.TransactionOpts{PreflightCommitment: rpc.CommitmentConfirmed})
	if err != nil {
		return solana.Signature{}, fmt.Errorf("failed to send transaction: %v", err)
	}

	return sig, nil
}
